lib/compiler.js

```diff
@@ -606,21 +607,51 @@ function resolveAppScriptPath(rootDir, relativePath, resolveOptions) {
   return (fixedFile === undefined ? resolvedPath : fixedFile);
 }
 
-function buildAllMixinInstructions(appRootDir, mixinDirs, options) {
+function buildAllMixinInstructions(appRootDir, mixinDirs, mixinSources, options,
+  modelInstructions) {
   var extensions = _.without(_.keys(require.extensions),
     _.keys(getExcludedExtensions()));
+
   var files = options.mixins || [];
+  var mixins = loadMixins(appRootDir, mixinDirs, files, extensions, options);
+  if (mixins === undefined) return;
+
+  /* If `mixinDirs` and `mixinSources` have any directories in common,
+   * then remove the common directories from `mixinSources` */
+  mixinSources = _.difference(mixinSources, mixinDirs);
+  files = [];
+  var mixinsFromMixinSources = loadMixins(appRootDir, mixinSources, files,
+    extensions, options);
+  if (mixinsFromMixinSources === undefined) return;
+
+  // Fetch unique list of mixin names, used in models
+  var modelMixins = _.unique(
+    fetchMixinNamesUsedInModelInstructions(modelInstructions));
+
+  // Filter-in only mixins, that are used in models
+  mixinsFromMixinSources = filterMixinInstructionsUsingWhitelist(
+    mixinsFromMixinSources, modelMixins);
+
+  mixins = _.reject(mixins, function(instruction) {
+    return (_.findWhere(mixinsFromMixinSources,
+      {name: instruction.name}) !== undefined);
+  });
+  mixins = mixins.concat(mixinsFromMixinSources);
+  return mixins;
+}
 
-  mixinDirs.forEach(function(dir) {
+function loadMixins(appRootDir, sourceDirs, files, extensions, options) {
+  sourceDirs.forEach(function(dir) {
     dir = tryResolveAppPath(appRootDir, dir);
     if (!dir) {
       debug('Skipping unknown module source dir %j', dir);
-      return;
+      return undefined;
```

Comment:
Please revert this line, the return value is ignored by `forEach` and `return` is the same as `return undefined`.